python/ray/_raylet.pyx

```diff
@@ -1518,6 +1529,203 @@ cdef class EmptyProfileEvent:
     def __exit__(self, *args):
         pass
 
+
+class RetryGcsConnectionError(Exception):
+    "If this error is raised, we should try to reconnect to GCS."
+    def __init__(self, exception):
+        self.exception = exception
+
+
+def _auto_reconnect(f):
+    @wraps(f)
+    def wrapper(self, *args, **kwargs):
+        if "TEST_RAY_COLLECT_KV_FREQUENCY" in os.environ:
+            ray._private.utils._CALLED_FREQ[f.__name__] += 1
+        remaining_retry = self._nums_reconnect_retry
+        while True:
+            try:
+                return f(self, *args, **kwargs)
+            except RetryGcsConnectionError as e:
+                if remaining_retry <= 0:
+                    raise RuntimeError(str(e.exception))
+                logger.debug(
+                    f"Failed to send request to gcs, reconnecting. Error {e.exception}"
+                )
+                try:
+                    self._connect()
+                except Exception:
+                    logger.error(f"Connecting to gcs failed. Error {e.exception}")
+                time.sleep(1)
+                remaining_retry -= 1
+                continue
+    return wrapper
+
+
+cdef class GcsClient:
```

Comment:
Do you happen to know the Python grpc client is this thread-safe by default? It'd be great we keep the same guarantee for thread safety.

Reply:
If we keep the GIL, the thread safety between the old pure python client and this client will be the same, if we drop the GIL we have to make sure the C++ implementation is thread safe, I can have a look at that.

Reply:
I looked more at this now and the APIs we use in the GcsSyncClient (namely, channels and stubs) are indeed thread safe https://github.com/grpc/grpc/issues/5649, so I think we are good here. It is not terribly well documented on the gRPC side though :|